lambdas/link_fetcher/app/subscription_endpoint.py

```diff
@@ -0,0 +1,211 @@
+        beginposition=iso8601.parse_date(payload["ContentDate"]["Start"]),
+        endposition=iso8601.parse_date(payload["ContentDate"]["End"]),
+        ingestiondate=iso8601.parse_date(payload["PublicationDate"]),
+        download_url=extracted["DownloadLink"],
+    )
+    return search_result
+
+
+def process_notification(
+    notification: dict[str, Any],
+    accepted_tile_ids: set[str],
+    session_maker: SessionMaker,
+    now_utc: Callable[[], datetime] = lambda: datetime.now(tz=timezone.utc),
+):
+    """Parse, filter, and potentially add new granule results to download queue"""
+    # Parse subscription notification to SearchResult
+    search_result = parse_search_result(notification["value"])
+
+    # Only consider imagery acquired in the last 30 days to avoid reprocessing of older imagery
+    oldest_acquisition_date = now_utc() - timedelta(days=30)
+    if search_result.beginposition < oldest_acquisition_date:
+        logger.info(f"Rejected {search_result=} (acquisition date too old)")
+        return
+
+    # Check tile ID
+    accepted_search_results = filter_search_results(
+        [search_result],
+        accepted_tile_ids,
+    )
+    if accepted_search_results:
+        logger.info(f"Adding {search_result=} to granule download queue")
+        add_search_results_to_db_and_sqs(
+            session_maker,
+            accepted_search_results,
+        )
+    else:
+        logger.info(f"Rejected {search_result=} (unacceptable tile)")
+
+
+def build_app(config: EndpointConfig) -> FastAPI:
+    """Create FastAPI app"""
+    app = FastAPI()
+    app.add_middleware(
+        CORSMiddleware,
+        allow_origins=["*"],
+        allow_credentials=True,
+        allow_methods=["*"],
+        allow_headers=["*"],
+    )
+    security = HTTPBasic()
+
+    accepted_tile_ids = get_accepted_tile_ids()
+
+    @app.post("/events", status_code=204)
+    def post_notification(
+        request: Request,
+        notification: dict[str, Any],
+        credentials: HTTPBasicCredentials = Depends(security),
+        session_maker: SessionMaker = Depends(get_session_maker),
+    ) -> Response:
+        """
+        Endpoint which uses Basic Auth and processes acquired notification.
+        """
+        # check Basic authorization
+        if not (
+            secrets.compare_digest(
+                credentials.username.encode("utf-8"),
+                config.notification_username.encode("utf-8"),
+            )
+            and secrets.compare_digest(
+                credentials.password.encode("utf-8"),
+                config.notification_password.encode("utf-8"),
+            )
+        ):
+            logging.error("Unauthorized")
+            raise HTTPException(status_code=401, detail="Unauthorized")
+
+        # process notification
+        process_notification(
+            notification=notification,
+            accepted_tile_ids=accepted_tile_ids,
+            session_maker=session_maker,
+        )
```

Comment:
```suggestion
        process_notification(
            notification=notification,
            accepted_tile_ids=accepted_tile_ids,
            session_maker=session_maker,
            now_utc=getattr(
                request.app, "now_utc", lambda: datetime.now(tz=timezone.utc)
            ),
         )
```

Reply:
Nice idea! Appreciate you pushing for this sort of thing. I ended up doing this a little differently, but still was able to inject the "what time is it?" dependency

Instead of storing the lookup function on the FastAPI application I added `now_utc` as a parameter to the `build_app` function. To me this felt like a little bit more direct than setting/getting an attribute off of the `.app` itself, and I think it mirrors how the `config` is injected into the FastAPI app builder

I could not remove `freezegun` as a dependency because a handful of tests related to the scheduled search-based link fetcher code still use it. This was the commit with the changes, https://github.com/NASA-IMPACT/hls-sentinel2-downloader-serverless/pull/44/commits/23ed64f9f27eccb512725ebe859ab66bc2b1cb99